### ISSUE TITLE
`execute` parallelization

### ISSUE THREAD

**Issue by LecrisUT** [teemtee/tmt#2755]
According to #976 `execute` is run with parallelization, but it is unclear how the parallelization works.

In my case I have MPI parallelization inside the tests, and I want to make sure that `tmt` does not parallelize on top of that, otherwise I get over-subscription and significant slow-down as compared to running the tmt tests sequentially. I couldn't find relevant documentation in either `plan.execute` or `test` specifications

**Comment by happz** [comment]
I'd say the term "parallel" alone is indeed puzzling and needs context to make sense. I can recommend https://tmt.readthedocs.io/en/stable/guide.html#multihost-testing, the second paragraph and the example below it. Yes, `execute` does perform something in parallel, but it's **phases** on **multiple guests**, not **multiple tests** on **single guest**. I assume you had just a single `discover` phase in your plan, which means its content - tests "discovered" by the said phase - were executed, in sequence, on the given guest. With `when` you can then instruct tmt to run the given phase on multiple guests at the same time:

**Comment by LecrisUT** [comment]
Ok that is reassuring, it's just the tests are dead slow :D

Hmm, what about setting multiple "hosts" that point to the same machine? That should correspond with running those in parallel, but probably that can be more efficiently by just letting `provision` handle that. One idea would be to support "multihost" testing on `provision.how=local`, but that would require some manipulations of cgroup and stuff to properly manage `processor` count and such.

**Comment by happz** [comment]
> Hmm, what about setting multiple "hosts" that point to the same machine? That should correspond with running those in parallel, but probably that can be more efficiently by just letting `provision` handle that. One idea would be to support "multihost" testing on `provision.how=local`, but that would require some manipulations of cgroup and stuff to properly manage `processor` count and such.

Note that this would run the same set of tests on the single guest multiple times - tmt does not know how to **split** the set of tests, run those sets on multiple guests (or just a single one but defined multiple times) and then merge their results like nothing happened. There should be an issue asking for some kind of `--max` option to allow partitioning.

**Comment by psss** [comment]
> In my case I have MPI parallelization inside the tests, and I want to make sure that `tmt` does not parallelize on top of that,

So, if I understand this correctly, your use case is covered as parallel execution of individual tests is not supported. As part of #2798 where parallel execution is proposed there should be definitely option to turn if off. My expectation would be to enable the parallel execution upon request.

@LecrisUT, is there anything still to be covered here or can we close the issue?

**Comment by LecrisUT** [comment]
Yes, it should be discussed there. Sorry for forgetting to follow on this one.
